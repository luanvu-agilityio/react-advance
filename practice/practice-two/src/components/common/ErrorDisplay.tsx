import styled from 'styled-components'
import { Theme, Button } from '@radix-ui/themes'
import { AlertTriangle, RefreshCw } from 'lucide-react'
import Text from './Text'

interface ErrorDisplayProps {
  error: string
  onRetry?: () => void
}

const ErrorContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  width: 100%;
  padding: 48px 32px;
  text-align: center;
  background-color: var(--red-1);
  border: 1px solid var(--red-4);
  border-radius: 12px;
  margin-top: 24px;
`

const ErrorDisplay = ({ error, onRetry }: ErrorDisplayProps) => (
  <ErrorContainer>
    <Theme>
      <AlertTriangle
        size={48}
        color="var(--red-9)"
        style={{ marginBottom: '16px' }}
      />
      <Text
        as="h3"
        text="Failed to Load Products"
        style={{
          fontSize: '20px',
          fontWeight: 'var(--font-weight-semibold)',
          color: 'var(--red-11)',
          margin: '0 0 8px 0',
        }}
      />
      <Text
        as="p"
        text={error}
        style={{
          fontSize: '16px',
          color: 'var(--red-10)',
          margin: '0 0 16px 0',
          maxWidth: '500px',
        }}
      />
      {onRetry && (
        <Button
          variant="solid"
          onClick={onRetry}
          style={{
            backgroundColor: 'var(--red-9)',
            color: 'white',
            display: 'flex',
            alignItems: 'center',
            gap: '8px',
          }}
        >
          <RefreshCw size={16} />
          Try Again
        </Button>
      )}
    </Theme>
  </ErrorContainer>
)

export default ErrorDisplay
