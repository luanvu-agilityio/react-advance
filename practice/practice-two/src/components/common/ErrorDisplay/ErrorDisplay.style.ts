import styled from 'styled-components'
import { Button } from '@radix-ui/themes'
import { AlertTriangle, RefreshCw } from 'lucide-react'

export const ErrorContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  width: 100%;
  padding: 48px 32px;
  text-align: center;
  background-color: var(--red-1);
  border: 1px solid var(--red-4);
  border-radius: 12px;
  margin-top: 24px;
`

export const StyledAlertIcon = styled(AlertTriangle)`
  margin-bottom: 16px;
  color: var(--coral-shade-1);
`

export const ErrorTitle = styled.h3`
  font-size: 20px;
  font-weight: var(--font-weight-semibold);
  color: var(--black-color-default);
  margin: 0 0 8px 0;
`

export const ErrorMessage = styled.p`
  font-size: 16px;
  color: var(--black-color-default);
  margin: 0 0 16px 0;
  max-width: 500px;
`

export const RetryButton = styled(Button)`
  background-color: var(--coral-color-default);
  color: white;
  font-size: 14px;
  font-weight: var(--font-weight-semibold);
  padding: 8px 16px;
  border-radius: 12px;
  display: flex;
  align-items: center;
  gap: 8px;

  &:hover {
    background-color: var(--coral-shade-2);
  }

  &:focus {
    box-shadow: 0 0 0 2px var(--coral-shade-1);
  }
`

export const RefreshIcon = styled(RefreshCw)`
  color: inherit;
`
