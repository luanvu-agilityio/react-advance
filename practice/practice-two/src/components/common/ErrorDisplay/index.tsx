import {
  ErrorContainer,
  StyledAlertIcon,
  ErrorTitle,
  ErrorMessage,
  RetryButton,
  RefreshIcon,
} from './ErrorDisplay.style'

export interface ErrorDisplayProps {
  error: string
  onRetry?: () => void
  title?: string
  retryText?: string
}

const ErrorDisplay = ({
  error,
  onRetry,
  title = 'Failed to Load Products',
  retryText = 'Try Again',
}: ErrorDisplayProps) => (
  <ErrorContainer>
    <StyledAlertIcon size={48} />

    <ErrorTitle>{title}</ErrorTitle>

    <ErrorMessage>{error}</ErrorMessage>

    {onRetry && (
      <RetryButton variant="solid" onClick={onRetry}>
        <RefreshIcon size={16} />
        {retryText}
      </RetryButton>
    )}
  </ErrorContainer>
)

export default ErrorDisplay
