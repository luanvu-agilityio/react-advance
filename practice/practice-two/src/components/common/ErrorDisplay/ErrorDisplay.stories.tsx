import type { Meta, StoryObj } from '@storybook/react'
import { action } from '@storybook/addon-actions'
import ErrorDisplay from './index'

const meta: Meta<typeof ErrorDisplay> = {
  title: 'Components/Common/ErrorDisplay',
  component: ErrorDisplay,
  parameters: {
    layout: 'centered',
    docs: {
      description: {
        component:
          'An error display component that shows error messages with an optional retry button. Used to handle error states across the application.',
      },
    },
  },
  argTypes: {
    error: {
      description: 'Error message to display',
      control: 'text',
    },
    onRetry: {
      description:
        'Optional retry function called when the retry button is clicked',
      action: 'onRetry',
    },
  },
  tags: ['autodocs'],
}

export default meta
type Story = StoryObj<typeof ErrorDisplay>

export const Default: Story = {
  args: {
    error: 'Something went wrong while loading the data.',
    onRetry: action('retry-clicked'),
  },
  parameters: {
    docs: {
      description: {
        story: 'Default error display with an error message and retry button.',
      },
    },
  },
}

export const WithoutRetryButton: Story = {
  args: {
    error: 'This error cannot be retried. Please contact support.',
  },
  parameters: {
    docs: {
      description: {
        story:
          'Error display without a retry button when onRetry is not provided.',
      },
    },
  },
}

export const NetworkError: Story = {
  args: {
    error:
      'Network connection failed. Please check your internet connection and try again.',
    onRetry: action('network-retry'),
  },
  parameters: {
    docs: {
      description: {
        story: 'Error display for network-related errors.',
      },
    },
  },
}

export const ServerError: Story = {
  args: {
    error:
      'Internal server error (500). Our team has been notified and is working on a fix.',
    onRetry: action('server-retry'),
  },
  parameters: {
    docs: {
      description: {
        story: 'Error display for server errors.',
      },
    },
  },
}

export const ValidationError: Story = {
  args: {
    error:
      'The data you submitted is invalid. Please check your input and try again.',
    onRetry: action('validation-retry'),
  },
  parameters: {
    docs: {
      description: {
        story: 'Error display for validation errors.',
      },
    },
  },
}

export const LongErrorMessage: Story = {
  args: {
    error:
      'This is a very long error message that demonstrates how the component handles extended text. It should wrap properly and maintain good readability even with multiple lines of error details. The component should handle this gracefully without breaking the layout.',
    onRetry: action('long-error-retry'),
  },
  parameters: {
    docs: {
      description: {
        story: 'Error display with a long error message to test text wrapping.',
      },
    },
  },
}

export const InContainer: Story = {
  render: (args) => (
    <div
      style={{
        width: '600px',
        height: '400px',
        border: '1px solid #e0e0e0',
        borderRadius: '8px',
        padding: '20px',
      }}
    >
      <ErrorDisplay {...args} />
    </div>
  ),
  args: {
    error: 'Failed to load user data from the server.',
    onRetry: action('container-retry'),
  },
  parameters: {
    docs: {
      description: {
        story:
          'Error display within a container to show how it fits in different layouts.',
      },
    },
  },
}

export const MultipleErrors: Story = {
  render: () => (
    <div
      style={{
        display: 'flex',
        flexDirection: 'column',
        gap: '20px',
        width: '800px',
      }}
    >
      <ErrorDisplay
        error="Failed to load user profile."
        onRetry={action('profile-retry')}
      />
      <ErrorDisplay
        error="Unable to fetch notifications."
        onRetry={action('notifications-retry')}
      />
      <ErrorDisplay error="Payment processing is currently unavailable." />
    </div>
  ),
  parameters: {
    docs: {
      description: {
        story: 'Multiple error displays to show consistent styling.',
      },
    },
  },
}
