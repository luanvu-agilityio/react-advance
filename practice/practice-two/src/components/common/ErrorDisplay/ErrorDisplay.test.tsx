import { render, screen, fireEvent } from '@testing-library/react'
import ErrorDisplay from './index'
import type { CSSProperties, ReactNode } from 'react'

// Mock the lucide-react icons
jest.mock('lucide-react', () => ({
  AlertTriangle: ({
    size,
    color,
    style,
  }: {
    size?: number | string
    color?: string
    style?: CSSProperties
  }) => (
    <div
      data-testid="alert-triangle-icon"
      data-size={size}
      data-color={color}
      style={style}
    >
      AlertTriangle
    </div>
  ),
  RefreshCw: ({ size }: { size?: number | string }) => (
    <div data-testid="refresh-icon" data-size={size}>
      RefreshCw
    </div>
  ),
}))

// Mock Radix UI components
jest.mock('@radix-ui/themes', () => ({
  Theme: ({ children }: { children: ReactNode }) => (
    <div data-testid="theme-wrapper">{children}</div>
  ),
  Button: ({
    children,
    onClick,
    variant,
    style,
  }: {
    children: ReactNode
    onClick?: () => void
    variant?: string
    style?: CSSProperties
  }) => (
    <button
      data-testid="retry-button"
      onClick={onClick}
      data-variant={variant}
      style={style}
    >
      {children}
    </button>
  ),
}))

// Mock the Text component
jest.mock('../Text/index', () => ({
  __esModule: true,
  default: ({
    text,
    as,
    style,
  }: {
    text: string
    as: string
    style: CSSProperties
  }) => (
    <div data-testid={`text-${as}`} data-as={as} style={style}>
      {text}
    </div>
  ),
}))

describe('ErrorDisplay', () => {
  const defaultProps = {
    error: 'Test error message',
  }

  // SNAPSHOT TESTS FOR RENDERING
  describe('Rendering', () => {
    it('renders all variants correctly', () => {
      // Default error display
      const { container: defaultContainer } = render(
        <ErrorDisplay {...defaultProps} />
      )
      expect(defaultContainer).toMatchSnapshot('default-error-display')

      // With retry button
      const onRetry = jest.fn()
      const { container: withRetryContainer } = render(
        <ErrorDisplay {...defaultProps} onRetry={onRetry} />
      )
      expect(withRetryContainer).toMatchSnapshot('with-retry-button')

      // With custom title
      const { container: customTitleContainer } = render(
        <ErrorDisplay {...defaultProps} title="Custom Error Title" />
      )
      expect(customTitleContainer).toMatchSnapshot('custom-title')

      // With custom retry text
      const { container: customRetryTextContainer } = render(
        <ErrorDisplay
          {...defaultProps}
          onRetry={() => {}}
          retryText="Custom Retry"
        />
      )
      expect(customRetryTextContainer).toMatchSnapshot('custom-retry-text')

      // With long error message
      const longError =
        'This is a very long error message that should be displayed properly without breaking the layout. It includes many words to ensure that text wrapping and other layout considerations are properly tested in the snapshot.'
      const { container: longErrorContainer } = render(
        <ErrorDisplay error={longError} />
      )
      expect(longErrorContainer).toMatchSnapshot('long-error-message')
    })
  })

  // INTERACTIVE BEHAVIOR TEST
  describe('Interactive behavior', () => {
    it('calls onRetry when retry button is clicked', () => {
      const onRetry = jest.fn()
      render(<ErrorDisplay {...defaultProps} onRetry={onRetry} />)

      const retryButton = screen.getByTestId('retry-button')
      fireEvent.click(retryButton)

      expect(onRetry).toHaveBeenCalledTimes(1)
    })
  })
})
