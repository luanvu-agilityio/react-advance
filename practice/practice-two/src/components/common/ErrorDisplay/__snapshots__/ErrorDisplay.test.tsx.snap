// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorDisplay Rendering renders all variants correctly: custom-retry-text 1`] = `
<div>
  <div
    class="sc-Qotzb eirdyb"
  >
    <div
      data-size="48"
      data-testid="alert-triangle-icon"
    >
      AlertTriangle
    </div>
    <h3
      class="sc-dKREkF hCKzTR"
    >
      Failed to Load Products
    </h3>
    <p
      class="sc-fWnslK imQEhB"
    >
      Test error message
    </p>
    <button
      data-testid="retry-button"
      data-variant="solid"
    >
      <div
        data-size="16"
        data-testid="refresh-icon"
      >
        RefreshCw
      </div>
      Custom Retry
    </button>
  </div>
</div>
`;

exports[`ErrorDisplay Rendering renders all variants correctly: custom-title 1`] = `
<div>
  <div
    class="sc-Qotzb eirdyb"
  >
    <div
      data-size="48"
      data-testid="alert-triangle-icon"
    >
      AlertTriangle
    </div>
    <h3
      class="sc-dKREkF hCKzTR"
    >
      Custom Error Title
    </h3>
    <p
      class="sc-fWnslK imQEhB"
    >
      Test error message
    </p>
  </div>
</div>
`;

exports[`ErrorDisplay Rendering renders all variants correctly: default-error-display 1`] = `
<div>
  <div
    class="sc-Qotzb eirdyb"
  >
    <div
      data-size="48"
      data-testid="alert-triangle-icon"
    >
      AlertTriangle
    </div>
    <h3
      class="sc-dKREkF hCKzTR"
    >
      Failed to Load Products
    </h3>
    <p
      class="sc-fWnslK imQEhB"
    >
      Test error message
    </p>
  </div>
</div>
`;

exports[`ErrorDisplay Rendering renders all variants correctly: long-error-message 1`] = `
<div>
  <div
    class="sc-Qotzb eirdyb"
  >
    <div
      data-size="48"
      data-testid="alert-triangle-icon"
    >
      AlertTriangle
    </div>
    <h3
      class="sc-dKREkF hCKzTR"
    >
      Failed to Load Products
    </h3>
    <p
      class="sc-fWnslK imQEhB"
    >
      This is a very long error message that should be displayed properly without breaking the layout. It includes many words to ensure that text wrapping and other layout considerations are properly tested in the snapshot.
    </p>
  </div>
</div>
`;

exports[`ErrorDisplay Rendering renders all variants correctly: with-retry-button 1`] = `
<div>
  <div
    class="sc-Qotzb eirdyb"
  >
    <div
      data-size="48"
      data-testid="alert-triangle-icon"
    >
      AlertTriangle
    </div>
    <h3
      class="sc-dKREkF hCKzTR"
    >
      Failed to Load Products
    </h3>
    <p
      class="sc-fWnslK imQEhB"
    >
      Test error message
    </p>
    <button
      data-testid="retry-button"
      data-variant="solid"
    >
      <div
        data-size="16"
        data-testid="refresh-icon"
      >
        RefreshCw
      </div>
      Try Again
    </button>
  </div>
</div>
`;
